File: examples/web/src/github-feedback/issueFormUrl.ts
export const MAX_QUERY_CHARS = 1800;

export type IssueFormFields = Record<string, string | undefined>;

export function isPlaceholderRepo(repo: string): boolean {
  const trimmed = repo.trim();
  return !trimmed || trimmed.toUpperCase() === "OWNER/REPO";
}

export function crashTitle(fingerprint: string, exceptionType: string): string {
  const fp = fingerprint
    .replace(/[^a-f0-9]/gi, "")
    .slice(0, 12)
    .toLowerCase();
  const kind = (exceptionType || "Error").split(/[^A-Za-z0-9_.$]/)[0] || "Error";
  return `[crash] ${fp} ${kind}`;
}

export function buildIssueFormUrl(
  repo: string,
  template: string,
  fields: IssueFormFields,
): { url: string; clipboardMarkdown?: string; bodyTooLarge: boolean } {
  const trimmed = repo.trim();
  if (isPlaceholderRepo(trimmed)) {
    return { url: "", bodyTooLarge: false };
  }
  const base = `https://github.com/${trimmed}/issues/new`;
  const params = new URLSearchParams();
  params.set("template", template);
  if (fields.title) params.set("title", fields.title);
  if (fields.labels) params.set("labels", fields.labels);
  for (const [key, value] of Object.entries(fields)) {
    if (!value || key === "title" || key === "labels") continue;
    params.set(key, value);
  }
  const full = `${base}?${params.toString()}`;
  if (full.length <= MAX_QUERY_CHARS) {
    return { url: full, bodyTooLarge: false };
  }
  const short = new URLSearchParams();
  short.set("template", template);
  if (fields.title) short.set("title", fields.title);
  const markdown =
    fields.stack || fields.description || fields.reproduction || fields.problem || "";
  return {
    url: `${base}?${short.toString()}`,
    clipboardMarkdown: markdown,
    bodyTooLarge: true,
  };
}

File: examples/android/app/src/main/java/dev/foss/goldenpath/githubfeedback/IssueFormUrl.kt
package dev.foss.goldenpath.githubfeedback

import java.net.URLEncoder
import java.nio.charset.StandardCharsets

object IssueFormUrl {
    const val MAX_QUERY_CHARS = 1800

    data class Built(
        val url: String,
        val clipboardMarkdown: String? = null,
        val bodyTooLarge: Boolean = false,
    )

    fun isPlaceholderRepo(repo: String): Boolean {
        val trimmed = repo.trim()
        return trimmed.isEmpty() || trimmed.equals("OWNER/REPO", ignoreCase = true)
    }

    fun crashTitle(fingerprint: String, exceptionType: String): String {
        val fp = fingerprint.filter { it.isLetterOrDigit() }.take(12).lowercase()
        val kind = exceptionType.split(Regex("[^A-Za-z0-9_.$]")).firstOrNull().orEmpty().ifEmpty { "Error" }
        return "[crash] $fp $kind"
    }

    fun build(repo: String, template: String, fields: Map<String, String>): Built {
        if (isPlaceholderRepo(repo)) return Built("")
        val base = "https://github.com/${repo.trim()}/issues/new"
        val params = linkedMapOf<String, String>()
        params["template"] = template
        fields["title"]?.let { params["title"] = it }
        fields["labels"]?.let { params["labels"] = it }
        fields.forEach { (key, value) ->
            if (key != "title" && key != "labels" && value.isNotEmpty()) params[key] = value
        }
        val full = base + "?" + encode(params)
        if (full.length <= MAX_QUERY_CHARS) return Built(full)
        val short = linkedMapOf("template" to template)
        fields["title"]?.let { short["title"] = it }
        val markdown = fields["stack"] ?: fields["description"] ?: fields["reproduction"] ?: fields["problem"] ?: ""
        return Built(base + "?" + encode(short), markdown, true)
    }

    private fun encode(params: Map<String, String>): String =
        params.entries.joinToString("&") { (k, v) ->
            enc(k) + "=" + enc(v)
        }

    private fun enc(value: String): String = URLEncoder.encode(value, StandardCharsets.UTF_8)
}
